File: backend/src/main/java/com/ufc/backend/backend/model/Course.java
package com.ufc.backend.backend.model;

import com.ufc.backend.backend.model.based.FeedbackBased;
import lombok.*;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.DBRef;
import org.springframework.data.mongodb.core.mapping.Document;
import java.util.List;
import java.util.Set;

@AllArgsConstructor
@NoArgsConstructor
@Getter
@Setter
@ToString
@Document
public class Course extends FeedbackBased {
    @Id
    private String id;
    private String name;
    private String shortName;
    @DBRef
    private Set<Classes> mandatoryClasses;
    @DBRef
    private Set<Classes> optionalClasses;

    public int hashCode() {
        return super.hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        if (getClass() != obj.getClass()) {
            return false;
        }
        return this.getId().equals(((Course) obj).getId());
    }
}

File: backend/src/main/java/com/ufc/backend/backend/model/Course.kt
package com.ufc.backend.backend.model

import com.ufc.backend.backend.commons.model.Identifiable
import org.hibernate.annotations.GenericGenerator
import org.springframework.data.annotation.Id
import javax.persistence.*

@Entity
data class Course
(
    @Id
    @GeneratedValue(generator = "uuid2")
    @GenericGenerator(name = "uuid2", strategy = "org.hibernate.id.UUIDGenerator")
    @Column
    override val id: String,
    @Column
    var name: String,
    @Column
    var shortName: String,
    @ManyToMany
    @JoinTable(
        name = "course_mandatory_subjects",
        joinColumns = [JoinColumn(name = "course_id")],
        inverseJoinColumns = [JoinColumn(name = "subject_id")]
    )
    var mandatoryClasses: Set<Subject>,
) : Identifiable<String> {

    override fun equals(other: Any?): Boolean {
        return if (javaClass != other!!.javaClass) {
            false
        } else this.id == (other as Course).id
    }

    override fun toString(): String {
        return "Course[id=$id, name=$shortName]"
    }

    override fun hashCode(): Int {
        return id.hashCode()
    }
}
